### PR TITLE
[Native DSL] Vendor quack library

### PR DESCRIPTION
**Pull request by slayton58**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.14.0) (oldest at bottom):
* #182319
* #182108
* __->__ #182107

Summary:

Vendor a subset of the quack library (https://github.com/Dao-AILab/quack)
under torch._vendor.quack.

Only the modules needed by the upcoming cuteDSL RMSNorm override are
vendored (rmsnorm.py) -- gemm/softmax/cross-entropy
and other unrelated paths are excluded for now. Imports within the vendored tree
are rewritten to be relative, and the ``quack::`` custom-op namespace is
renamed to ``torch_vendor_quack::``, so this copy is fully independent of any
``quack`` package installed via pip.

Note: Pyrefly disabled for this new directory

Added tooling to generate the vendored subset (via. claude):

```
tools/vendoring/quack/vendor.sh <sha>                    # clones upstream
tools/vendoring/quack/vendor.sh <sha> <local-checkout>   # reuses a clone
```

This pulls upstream at the provided hash, strips out the necessary
files, and applies the included patchset. This should a) make updating
the vendored files easier in the future and b) be extendible later on if
we want to add more functionality - add to the list of files to be
brought in, and create a new patch-set for those files. Also makes sure
upstream `LICENSE` file is copied & copyright/attribution between
upstream & vendored is bitwise identical.

Upstream SHA: 6bceaad2dba3b979b898824b146b1bb2816fc483 (quack 0.4.0)

Signed-off-by: Simon Layton <simonlayton@meta.com>

cc @albanD

### CODE REVIEW

File: torch/_vendor/quack/__init__.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+
+if not any(isinstance(f, _QuackAliasFinder) for f in sys.meta_path):
+    # Insert before PathFinder so torch._vendor.quack.X is resolved by us,
+    # not by Python's path-based import (which would otherwise find quack's
+    # source via the __path__ inherited from the quack module and execute
+    # submodules a second time under a different name).
+    sys.meta_path.insert(0, _QuackAliasFinder())
+
+import quack as _quack  # noqa: E402
```
Comment (slayton58) [inline torch/_vendor/quack/__init__.py]: I realize there needs to be a guard against quack installed regularly / explicitly a relative import here

File: torch/_vendor/quack/cache_utils.py [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+CACHE_DIR: str | None = os.getenv("QUACK_CACHE_DIR", None)
+COMPILE_ONLY: bool = False
+
+# Downstream projects can append directories here to include their sources
+# in the cache fingerprint. Must be set before the first jit_cache call.
+EXTRA_SOURCE_DIRS: list[Path] = []
+
+EXPORT_FUNC_NAME = "func"
+LOCK_TIMEOUT = 60
+CacheInfo = namedtuple("CacheInfo", ["hits", "misses", "maxsize", "currsize"])
```
Comment (Skylion007) [inline torch/_vendor/quack/cache_utils.py:39]: This NamedTuple API is deprecated use typed NamedTuple API

File: torch/_vendor/quack/cache_utils.py [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+        deadline = time.monotonic() + self.timeout
+        while time.monotonic() < deadline:
+            try:
+                fcntl.flock(self._fd, lock_type | fcntl.LOCK_NB)
+                return self
+            except OSError:
+                time.sleep(0.1)
+        os.close(self._fd)
+        self._fd = -1
+        raise RuntimeError(f"Timed out waiting for lock: {self.lock_path}")
```
Comment (Skylion007) [inline torch/_vendor/quack/cache_utils.py:110]: Doesn't Python have a builtin TimeoutException? Raise that.

File: torch/_vendor/quack/cache_utils.py [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+            os.close(self._fd)
+            self._fd = -1
+
+
+# ---------------------------------------------------------------------------
+# JIT cache decorator
+# ---------------------------------------------------------------------------
+
+
+def jit_cache(fn):
```
Comment (Skylion007) [inline torch/_vendor/quack/cache_utils.py:124]: Use Callable + ParamSpec + TypeVar to preserve typing.

File: torch/_vendor/quack/cache_utils.py [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+            lock_path = cache_path / f"{sha}.lock"
+            try:
+                with FileLock(lock_path, exclusive=False, timeout=LOCK_TIMEOUT):
+                    if o_path.exists():
+                        m = cute.runtime.load_module(str(o_path), enable_tvm_ffi=True)
+                        loaded = m[EXPORT_FUNC_NAME]
+                        cache[cache_key] = loaded
+                        hits += 1
+                        return _noop_kernel if COMPILE_ONLY else loaded
+            except RuntimeError:
```
Comment (Skylion007) [inline torch/_vendor/quack/cache_utils.py:160]: TimeoutError

File: torch/_vendor/quack/cache_utils.py [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+            try:
+                with FileLock(lock_path, exclusive=True, timeout=LOCK_TIMEOUT):
+                    if not o_path.exists():
+                        o_path.parent.mkdir(parents=True, exist_ok=True)
+                        compiled_fn.export_to_c(
+                            object_file_path=str(o_path),
+                            function_name=EXPORT_FUNC_NAME,
+                        )
+            except Exception as e:
+                print(f"torch._vendor.quack cache: export failed for key {sha}: {e}")
```
Comment (Skylion007) [inline torch/_vendor/quack/cache_utils.py:179]: Use logger?

**Review by Skylion007 — COMMENTED** [review]
NVM, I was reviewing the vendored code.

**Review by albanD — APPROVED** [review]
Sounds ok.
Please update the .md for torch/_vendor with the appropriate details.
Also please add yourself as a codeowner for this folder so you can monitor if anyone tries to modify the code not as a patch, but directly inside.

**Comment by slayton58** [comment]
> Sounds ok. Please update the .md for torch/_vendor with the appropriate details. Also please add yourself as a codeowner for this folder so you can monitor if anyone tries to modify the code not as a patch, but directly inside.

Good points - done
